### ISSUE TITLE
optimize - check that maxiter is not exceeded.

### ISSUE THREAD

**Issue by andyfaff** [scipy/scipy#8375]
At the moment it's possible to set a `maxiter` parameter for many optimizers. Up until now the behaviour for many of them (e.g. #7854) has been to stop when `n_iterations > maxiter` (typically `maxiter + 1`. This is a bit weird because you always end up running one too many iterations. Sometimes the OptimizeResult is not the same as `n_iterations == maxiter`, but `n_iterations == maxiter + 1`. I propose that `maxiter` is enforced more rigourously.

i.e.

- [ ] add a test for all `minimize` methods to check that maxiter is not exceeded.
- [ ] add a test to check that OptimizeResult.x, OptimizeResulf.fun are the same as the last callback (which should be at the end of the iteration).
- [ ] change the `maxiter` [error message](https://github.com/scipy/scipy/blob/master/scipy/optimize/optimize.py#L47) to "been reached".
- [ ] fix optimizers as required for the tests to pass.

This is being more strict, and apart from the error message changing I can't any backcompat issues.

**Comment by josef-pkt** [comment]
I agree with it as a bugfix, but it is still not fully backwards compatibile. E.g. downstream unit tests might fail if values are tested when maxiter is binding.

E.g. IIRC, we had a case in statsmodels where maxiter was (accidentially) set to the same number as convergence. This failed across some version when convergence was a bit slower or maxiter behavior had changed.

If `n_iterations == maxiter + 1` is consistent across most optimizers and is documented (see error message), then I think it doesn't need to be changed. maybe just more clearly documented.

(Aside: in statsmodels we also use maxiter=0 to just run through the function or method at the starting values, but it's not consistent yet across models. It might be one optimization step if we don't catch it.)

**Comment by brandondube** [comment]
I do not think scipy should defer positive changes because they might break backwards compatibility.  The fact of the matter is that as a user, when I enter maxiter=5, I fully expect 5 or fewer iterations.  Not 6, not 7.  5.

I also expect the callback to be called exactly 5 times, not 6.  Even more so, I expect an equal number of callback calls to function iterations.

**Comment by andyfaff** [comment]
@josef-pkt FWIW the documentation for `scipy.optimize.minimize` is:

>A dictionary of solver options. All methods accept the following generic options:
maxiter : int
Maximum number of iterations to perform.

**Comment by andyfaff** [comment]
>  E.g. downstream unit tests might fail if values are tested when maxiter is binding.

A problem in #7854 was that `maxiter + 2` iterations were occurring, yet the last callback received an `x` from `maxiter + 1`, and the final `x` was different again.

**Comment by jaimefrio** [comment]
I found about and reported #7854 because we use optimize at work, and when running it from a Jupyter notebook we display a progress bar that gets updated by the callback, and if the bar exceeded 100% something else broke in our code. We use several other methods, and I only saw this happening with l-bfgs-b. I will double check tomorrow, but I'm pretty confident that bfgs, cg and newton-cg call the callback at most the max number of iterations.

But it would indeed be nice if we didn't have to depend on me being confident, but on the tests checking it and passing!

**Comment by brandondube** [comment]
Some additional info:
for n < 24, maxiter+2 iterations will be done for L-BFGS-B.  For n=24, maxiter+1 iterations will be done.  For maxiter>=25, maxiter iterations will be done.

In the cases where maxiter +2 or maxiter +1 iterations are done, the callback is called a different number of times to the number of iterations done.

**Comment by andyfaff** [comment]
Lbfgs has-been fixed.
